### PR TITLE
feat(events): adds event emitter instrumentation

### PR DESCRIPTION
**Pull request by kevzettler**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -110,12 +115,13 @@ function memoizeFn(client, options, lock, fn,
     // Ok, we're going to have to actually execute the function.
     // Lock ensures only one fn executes at a time and prevents a stampede.
     const unlock = await lock(key, lock_timeout);
+    if(unlock === null) options.emitter(`lock_timeout.${key}`);
```
Comment (STRML) [inline index.js]: Missing `emit`

Comment (STRML) [reply index.js]: Didn't actually realize unlock could be `null` - we should fix the `finally` so it doesn't crash.

File: lock.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ const {exec} = require('./redisCompat');
 async function acquireLock(client, lockName, timeoutStamp, retryDelay) {
   try {
     const timeoutLeft = timeoutStamp - Date.now();
-    if (timeoutLeft <= 0) return;
+    if (timeoutLeft <= 0) return null;
```
Comment (STRML) [inline lock.js]: This doesn't change anything, see https://github.com/BitMEX/redis-memoizer/blob/bc5fa2b45579b15ba762df2ff6218abbbc46f941/lock.js#L33

File: index.js [hunk 3]
```diff
@@ -72,6 +74,9 @@ function createMemoizeFunction(client, options = {}) {
     if (typeof options.on_error !== 'function') {
       throw new Error('An error fn of the arity (err, client, key) must be passed as an option.');
     }
+    if(!(options.emitter instanceof EventEmitter)){
+      throw new Error('An emitterd passed to the memoizer must be an instance of EventEmitter');
```
Comment (sandbochs) [inline index.js]: typo

File: index.js [hunk 4]
```diff
@@ -115,7 +120,7 @@ function memoizeFn(client, options, lock, fn,
       // After we've acquired the lock, check if the key was populated in the meantime.
       const memoValueRetry = await doLookup(client, key, timeoutMs, options);
       if (memoValueRetry !== MAGIC.not_found) return memoValueRetry;
-
+      options.emitter.emit('miss', key);
```
Comment (sandbochs) [inline index.js]: can the event names be more descriptive or is it obvious that this is a redis cache miss?

File: index.js [hunk 5]
```diff
@@ -140,18 +145,24 @@ function memoizeFn(client, options, lock, fn,
...
   try {
     memoValue = await Promise.resolve(getKeyFromRedis(client, key, options)).timeout(timeout);
+    options.emitter.emit('hit', key);
   } catch (err) {
     err.message = `Redis-Memoizer: Error getting key "${key}" with timeout ${timeout}: ${err.message}`;
-    if (err.name !== 'TimeoutError') {
+    if (err.name === 'TimeoutError') {
+      options.emitter.emit('lookup_timeout', key);
+    }else{
       options.on_error(err, client, key);
```
Comment (sandbochs) [inline index.js]: do we want to emit generic errors?

File: lock.js [hunk 6]
```diff
@@ -5,10 +5,13 @@ const {exec} = require('./redisCompat');
 // Variant of redis-lock intended for use with redis-memoizer. Unlike redis-lock,
 // this instead takes an overall timeout, after which the lock is disregarded. This ensures
 // that functions just carry on if the lock is held for too long.
-async function acquireLock(client, lockName, timeoutStamp, retryDelay) {
+async function acquireLock(client, lockName, timeoutStamp, retryDelay, emitter) {
   try {
     const timeoutLeft = timeoutStamp - Date.now();
-    if (timeoutLeft <= 0) return;
+    if (timeoutLeft <= 0) {
+      if(emitter) emitter.emit('lock_timeout', lockName);
```
Comment (sandbochs) [inline lock.js]: is the snake case a pattern in the event aggregator?

File: index.js (hunk 4 above)
Comment (kevzettler) [reply index.js]: See https://github.com/BitMEX/api/compare/feat/addl-statsd?expand=1#diff-5027b6db5bd65c60e6aa7906d85197ccR24

For and example of how these get collected

File: index.js (hunk 5 above)
Comment (sandbochs) [reply index.js]: @kevzettler not sure why this comment got hidden

File: lock.js (hunk 6 above)
Comment (kevzettler) [reply lock.js]: updated to camelCase

File: index.js (hunk 5 above)
Comment (kevzettler) [reply index.js]: Yeah me either, good question. cc @STRML

Comment (STRML) [reply index.js]: Yeah why not.

**Comment by kevzettler** [comment]
@STRML this is ready for another pass

**Pull request merged**